src/data/padics/ring_homs.lean

```diff
@@ -0,0 +1,701 @@
+satisfying `∥(x - zmod_repr x : ℤ_[p])∥ < 1`.
+-/
+def zmod_repr : ℕ :=
+classical.some (exists_mem_range x)
+
+lemma zmod_repr_spec : zmod_repr x < p ∧ (x - zmod_repr x ∈ maximal_ideal ℤ_[p]) :=
+classical.some_spec (exists_mem_range x)
+
+lemma zmod_repr_lt_p : zmod_repr x < p := (zmod_repr_spec _).1
+
+lemma sub_zmod_repr_mem : (x - zmod_repr x ∈ maximal_ideal ℤ_[p]) := (zmod_repr_spec _).2
+
+/--
+`to_zmod_hom` is an auxiliary constructor for creating ring homs from `ℤ_[p]` to `zmod v`.
+-/
+def to_zmod_hom (v : ℕ) (f : ℤ_[p] → ℕ) (f_spec : ∀ x, x - f x ∈ (ideal.span {v} : ideal ℤ_[p]))
+  (f_congr : ∀ (x : ℤ_[p]) (a b : ℕ),
+     x - a ∈ (ideal.span {v} : ideal ℤ_[p]) → x - b ∈ (ideal.span {v} : ideal ℤ_[p]) →
+       (a : zmod v) = b) :
+  ℤ_[p] →+* zmod v :=
+{ to_fun := λ x, f x,
+  map_zero' :=
+  begin
+    rw [f_congr (0 : ℤ_[p]) _ 0, cast_zero],
+    { exact f_spec _ },
+    { simp only [sub_zero, cast_zero, submodule.zero_mem], }
+  end,
+  map_one' :=
+  begin
+    rw [f_congr (1 : ℤ_[p]) _ 1, cast_one],
+    { exact f_spec _ },
+    { simp only [sub_self, cast_one, submodule.zero_mem], }
+  end,
+  map_add' :=
+  begin
+    intros x y,
+    rw [f_congr (x + y) _ (f x + f y), cast_add],
+    { exact f_spec _ },
+    { convert ideal.add_mem _ (f_spec x) (f_spec y),
+      rw cast_add,
+      ring, }
+  end,
+  map_mul' :=
+  begin
+    intros x y,
+    rw [f_congr (x * y) _ (f x * f y), cast_mul],
+    { exact f_spec _ },
+    { let I : ideal ℤ_[p] := ideal.span {v},
+      have A : x * (y - f y) ∈ I := I.mul_mem_left (f_spec _),
+      have B : (x - f x) * (f y) ∈ I := I.mul_mem_right (f_spec _),
+      convert I.add_mem A B,
+      rw cast_mul,
+      ring, }
+  end, }
+
+/--
+`to_zmod` is a ring hom from `ℤ_[p]` to `zmod p`,
+with the equality `to_zmod x = (zmod_repr x : zmod p)`.
+-/
+def to_zmod : ℤ_[p] →+* zmod p :=
+to_zmod_hom p zmod_repr
+  (by { rw ←maximal_ideal_eq_span_p, exact sub_zmod_repr_mem })
+  (by { rw ←maximal_ideal_eq_span_p, exact zmod_congr_of_sub_mem_max_ideal } )
+
+/--
+`z - (to_zmod z : ℤ_[p])` is contained in the maximal ideal of `ℤ_[p]`, for every `z : ℤ_[p]`.
+
+The coercion from `zmod p` to `ℤ_[p]` is `zmod.has_coe_t`,
+which coerces `zmod p` into artibrary rings.
+This is unfortunate, but a consequence of the fact that we allow `zmod p`
+to coerce to rings of arbitrary characteristic, instead of only rings of characteristic `p`.
+This coercion is only a ring homomorphism if it coerces into a ring whose characteristic divides `p`.
+While this is not the case here we can still make use of the coercion.
+-/
+lemma to_zmod_spec (z : ℤ_[p]) : z - (to_zmod z : ℤ_[p]) ∈ maximal_ideal ℤ_[p] :=
+begin
+  convert sub_zmod_repr_mem z using 2,
+  dsimp [to_zmod, to_zmod_hom],
+  unfreezingI { rcases (exists_eq_add_of_lt (hp_prime.pos)) with ⟨p', rfl⟩ },
+  change ↑(zmod.val _) = _,
+  simp [zmod.val_cast_nat],
+  rw mod_eq_of_lt,
+  simpa only [zero_add] using zmod_repr_lt_p z,
```

Comment:
These tactics look quite fragile. Could you pass all these arguments to `simpa`, or use `simp only`/`simp_rw` instead?.

Reply:
The `rw` step is actually solving one goal and creating another, and is really an `apply` in disguise. I've removed the nonterminal simp and changed `rw` to `apply`.